Add a filter to `build_query_vars_from_query_block`

sunyatasattva opened the pull request:
## What?
This PR adds a filter to the `build_query_vars_from_query_block` which is used to parse a `WP_Query` and pass it down the Query Loop block context. 

## Why?
This allows Query Loop block variations to hook into the query and possibly inject custom query parameters, enabling developers to support all kinds of custom queries and create powerful variations on the Query Loop block.

For an example on how this can be used to achieve this goal, see this PR: https://github.com/woocommerce/woocommerce-blocks/pull/6812 (specifically, this filter is used [here](https://github.com/woocommerce/woocommerce-blocks/pull/6812/files#diff-efb947e345af5e501dfaa2980b055b3b31c308b31e8db23049b6b48fecee33d8R39-R52)).

Paging @gziolo and @ntsekouras.

## How?

The minimal change to enable this is to wrap the return value of the function in a filter.

## Testing Instructions

1. Add a callback to your PHP code that hooks to this filter, e.g.

```
add_filter(
  'gutenberg_build_query_vars_from_query_block',
  function ( $query ) {
    $query[ 'offset' ] = 1;

    return $query;
 },
 10,
 1
);
```

2. Add multiple Query Loop blocks to your pages with different settings.
3. Save the page and go to the front-end.
4. Notice that all the blocks will render with your modified query (e.g. start from an offset of one, if the above is used).

A way to test the practical use of this change would be to use the [above-mentioned PR](https://github.com/woocommerce/woocommerce-blocks/pull/6812) to see how it could work. Or just pull from `trunk` on [WooCommerce Blocks](https://github.com/woocommerce/woocommerce-blocks).

sunyatasattva committed:
Add a filter to `build_query_vars_from_query_block`

The filter allows Query Loop block variation to hook into the query
and possibly inject custom query parameters.

gziolo commented:
@sunyatasattva, thank you for opening this PR. I see this filter applicable in many more cases. It would help to handle custom post types or even custom fields for existing REST API endpoints registered by plugins 👍🏻 

The way to move forward with this PR would be to include the documentation for the hook like here in this repo:

https://github.com/WordPress/gutenberg/blob/79dcf05b6a93798272b9eed5c778cddb49b5dd32/packages/block-library/src/navigation/index.php#L334-L345

or like here in WordPress core:

https://github.com/WordPress/wordpress-develop/blob/7f7d616d822b79c952cbd6a3046a6f6a8aa5a35e/src/wp-includes/blocks.php#L262-L269

We also need to settle on the name of the hook that is going to stay forever. It can't be prefixed with `gutenberg_` and it can be shorter than the function name. You can check the existing WordPress hooks for inspiration at https://developer.wordpress.org/reference/hooks/.

We will also need to move the updated function from `lib/compat/wordpress-6.0/blocks.php` to `lib/compat/wordpress-6.1/blocks.php` to ensure it polyfills the functionality in the Gutenberg plugin correctly. Later, we will need to backport those changes to WordPress Core to land them in WordPress 6.1 (planned for November 1st).

ntsekouras commented:
Thanks for the PR @sunyatasattva! Everything @gziolo said is great 😄 

We also need to look into the editor side(related: https://github.com/WordPress/gutenberg/issues/34201), but it can happen in a different PR.

spacedmonkey commented:
Php docs are a must have.

sunyatasattva committed:
Move `build_query_vars_from_query_block` to 6.1 compat and add docs

sunyatasattva commented:
Hello @gziolo thank you for your review. I pushed another commit, let's see if I've understood correctly:

1. The function has been completely moved from `compat/wordpress-6.0` to `compat/wordpress-6.1`, not just duplicated with the new changes. Is that right?
2. I added the docs to the filter.
3. I've changed the name of the filter to `query_block_query_vars`.

The reasoning for my proposal on that name is that:

1. Filters are not named like verbs, but rather should be nouns.
2. They should represent what is output by the filter itself.
3. There are such names for other filters as `query_vars` or `removable_query_args`. It seemed to me that such a name could match those naming conventions.

Let me know what you think and if I did everything right.

ntsekouras commented on lib/compat/wordpress-6.1/blocks.php:
```diff
@@ -205,3 +205,139 @@ function gutenberg_block_type_metadata_multiple_view_scripts( $metadata ) {
+		if ( ! empty( $block->context['query']['search'] ) ) {
+			$query['s'] = $block->context['query']['search'];
+		}
+		if ( ! empty( $block->context['query']['parents'] ) && is_post_type_hierarchical( $query['post_type'] ) ) {
```
Interestingly enough this code from [here](https://github.com/WordPress/gutenberg/pull/40933) was close to the 6.0 release and didn't land there, so we should have moved this file to `compat/6.1` either way.. Good timing for this change 😄

ntsekouras commented on lib/compat/wordpress-6.1/blocks.php:
```diff
@@ -205,3 +205,139 @@ function gutenberg_block_type_metadata_multiple_view_scripts( $metadata ) {
+ *
+ * It's used in QueryLoop, QueryPaginationNumbers and QueryPaginationNext blocks.
+ *
+ * `build_query_vars_from_query_block` was introduced in 5.8, for 6.0 we just need
```
```suggestion
 * `build_query_vars_from_query_block` was introduced in 5.8, for 6.1 we just need
```

ntsekouras commented on lib/compat/wordpress-6.1/blocks.php:
```diff
@@ -205,3 +205,139 @@ function gutenberg_block_type_metadata_multiple_view_scripts( $metadata ) {
+	 *
+	 * @param array    $query Array containing parameters for `WP_Query` as parsed by the block context.
+	 * @param WP_Block $block Block instance.
+	 * @param int      $page  Current query page.
```
```suggestion
	 * @param int      $page  Current query's page.
```

ntsekouras commented on lib/compat/wordpress-6.1/blocks.php:
```diff
@@ -205,3 +205,139 @@ function gutenberg_block_type_metadata_multiple_view_scripts( $metadata ) {
+	 *
+	 * Anything to this filter should be compatible with the `WP_Query` API to form
+	 * the query context which will be passed down to the Query Loop Block's children.
+	 * This can help, for example, to include additional custom post types or meta queries not
```
I'd avoid mentioning `custom post types` as they already are supported by the block.

ntsekouras commented on lib/compat/wordpress-6.1/blocks.php:
```diff
@@ -205,3 +205,139 @@ function gutenberg_block_type_metadata_multiple_view_scripts( $metadata ) {
+	/**
+	 * Filters the arguments which will be passed to `WP_Query` for the Query Loop Block
+	 *
+	 * Anything to this filter should be compatible with the `WP_Query` API to form
```
Should we add to the comments that while this will be possible with this filter server side, we would still need a way to have the same effect in the editor and that is not the case for now? 🤔 

While I don't think it's bad that some third party devs could choose to have different results client/server side, it should be made clear that in client side we use the REST API and it doesn't support all the arguments WP_Query does.

Ideally devs should aim to provide arguments that are also supported in REST API, so as to have identical queries - when we implement that client side.

sunyatasattva replied:
Yep, totally! Will add this.

ntsekouras commented on lib/compat/wordpress-6.1/blocks.php:
```diff
@@ -205,3 +205,139 @@ function gutenberg_block_type_metadata_multiple_view_scripts( $metadata ) {
+	 * @param WP_Block $block Block instance.
+	 * @param int      $page  Current query page.
+	 */
+	return apply_filters( 'query_block_query_vars', $query, $block, $page );
```
Maybe
```suggestion
	return apply_filters( 'query_loop_block_query_vars', $query, $block, $page );
```

ntsekouras reviewed (commented):
Thanks for iterating @sunyatasattva! 💯

sunyatasattva committed:
Apply suggestions from code review

* Change filter name to `query_loop_block_query_vars`
* Fix docs typo
* Fix docs version

Co-authored-by: Nik Tsekouras <ntsekouras@outlook.com>

sunyatasattva committed:
Elaborate filter docs

sunyatasattva commented:
@ntsekouras Feedback implemented, let me know if there is anything else that should be done 🙏

ntsekouras commented on lib/compat/wordpress-6.1/blocks.php:
```diff
@@ -205,3 +205,145 @@ function gutenberg_block_type_metadata_multiple_view_scripts( $metadata ) {
+	}
+
+	/**
+	 * Filters the arguments which will be passed to `WP_Query` for the Query Loop Block
```
```suggestion
	 * Filters the arguments which will be passed to `WP_Query` for the Query Loop Block.
```

ntsekouras commented on lib/compat/wordpress-6.1/blocks.php:
```diff
@@ -205,3 +205,145 @@ function gutenberg_block_type_metadata_multiple_view_scripts( $metadata ) {
+	 *
+	 * Anything to this filter should be compatible with the `WP_Query` API to form
+	 * the query context which will be passed down to the Query Loop Block's children.
+	 * This can help, for example, to include additional settings or meta queries not
+	 * directly supported by the core Query Loop Block, and extend its capabilities.
```
I think this is the key part of the comment. Anything already supported by the block should be done by the block. For example we have special handling for `offset` based the `perPage and page` attribute, and if someone would alter only the `perPage` from the filter, would end up with wrong offset in the pagination..

sunyatasattva replied:
Yea, I see that. I guess that's the risk with many filters, right? Each developer needs to be responsible and see what kind of side effects they might be introducing…

ntsekouras reviewed (approved):
Great work, thanks @sunyatasattva! This looks good to me! 💯 

I'd appreciate a final look from @gziolo, but I think it's good to 🚢

ntsekouras committed:
Update lib/compat/wordpress-6.1/blocks.php

gziolo commented:
Awesome, thank you @sunyatasattva for bringing it to the finish line. Once, we have also the matching API in the editor, let's make sure it is documented in the Block Editor Handbook somewhere in the hooks section:

https://github.com/WordPress/gutenberg/tree/trunk/docs/reference-guides/filters

gziolo commented:
@sunyatasattva, can you resolve conflicts in this branch. It looks like I landed changes added in exactly the same place and git has some trouble understanding those are different features and both should be there 😅

sunyatasattva committed:
Merge branch 'trunk' into add/build-query-vars-from-query-block-filter

gziolo commented on lib/compat/wordpress-6.1/blocks.php:
```diff
@@ -207,6 +207,148 @@ function gutenberg_block_type_metadata_multiple_view_scripts( $metadata ) {
+	return apply_filters( 'query_loop_block_query_vars', $query, $block, $page );
+}
+
+/*
```
It looks like my mistake from another PR 😅 
```suggestion
/**
```

Fixes:

<img width="694" alt="Screen Shot 2022-08-31 at 12 15 28" src="https://user-images.githubusercontent.com/699132/187656169-d5a5dcb6-ebf4-4bab-983b-6c2f097836e9.png">

sunyatasattva replied:
Thank you @gziolo !

gziolo committed:
Update lib/compat/wordpress-6.1/blocks.php
